File: src/main/java/guru/drinkit/service/impl/BasicUserDetailsService.java
package guru.drinkit.service.impl;

import java.util.Collection;
import javax.annotation.Resource;

import guru.drinkit.common.DetailedUser;
import guru.drinkit.common.Role;
import guru.drinkit.domain.User;
import guru.drinkit.repository.UserRepository;
import org.apache.commons.collections4.Transformer;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Component;

import static org.apache.commons.collections4.CollectionUtils.collect;

@Component
public class BasicUserDetailsService implements UserDetailsService {

    @Resource
    private UserRepository userRepository;

    @Override
    public UserDetails loadUserByUsername(String username) {
        User user = userRepository.findByUsername(username);
        if (user == null) {
            throw new UsernameNotFoundException("Test msg");
        }
        return new DetailedUser(user.getUsername(), user.getPassword(),
                getAuthorities(user.getAccessLevel()), user.getDisplayName());
    }

    private Collection<? extends GrantedAuthority> getAuthorities(Integer accessLevel) {
        return collect(Role.getRolesByAccessLevel(accessLevel), new Transformer<Role, SimpleGrantedAuthority>() {
            @Override
            public SimpleGrantedAuthority transform(final Role role) {
                return new SimpleGrantedAuthority(role.name());
            }
        });
    }

    public boolean createUser(User user) {
        boolean isNew = userRepository.findByUsername(user.getUsername()) == null;
        if(isNew){
            userRepository.save(user);
        }
        return isNew;
    }


}

File: src/main/java/guru/drinkit/service/impl/BasicUserDetailsService.kt
package guru.drinkit.service.impl

import guru.drinkit.common.ExtendedUser
import guru.drinkit.domain.User
import guru.drinkit.repository.UserRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.stereotype.Component

@Component
class BasicUserDetailsService @Autowired constructor(
        private val userRepository: UserRepository
) : UserDetailsService {


    override fun loadUserByUsername(username: String): UserDetails {
        val user = userRepository.findByUsername(username) ?: throw UsernameNotFoundException("Test msg")
        return ExtendedUser(user)
    }

    fun createUser(user: User): Boolean {
        val isNew = userRepository.findByUsername(user.username) == null
        if (isNew) {
            userRepository.save(user)
        }
        return isNew
    }


}
